### DISCUSSION TITLE
How to compile in android

### DISCUSSION THREAD

**Discussion by mohammada315** [mavlink-router/mavlink-router#356] [Q&A]
Dear Mavlink Developer
I need to run this project on Android like HereLink 

Thanks for your cooperation

**Comment by ashwinr-act-cit** [comment]
Hi, mavlink-router doesn't have official Android support out of the box, but you can cross-compile it for Android using the Android NDK. Here's how:

## Prerequisites

1. **Android NDK** - Download from Android Studio (Tools > SDK Manager > SDK Tools > NDK)
2. **CMake** - Also available through Android Studio SDK Manager
3. **Meson** - The build system used by mavlink-router (install via `pip3 install meson`)

## Build Steps

```bash
# Clone the repo with submodules
git clone --recursive https://github.com/mavlink-router/mavlink-router
cd mavlink-router

# Set up NDK environment
export ANDROID_NDK_HOME=/path/to/android/ndk
export ANDROID_PLATFORM=android-28
export ANDROID_ABI=arm64-v8a  # or armeabi-v7a for 32-bit

# Configure with Meson using Android toolchain
meson setup build \
  -Dc_args="--sysroot=$ANDROID_NDK_HOME/toolchains/llvm/prebuilt/linux-x86_64/sysroot" \
  -Dcpp_args="--sysroot=$ANDROID_NDK_HOME/toolchains/llvm/prebuilt/linux-x86_64/sysroot" \
  -Dc_link_args="--sysroot=$ANDROID_NDK_HOME/toolchains/llvm/prebuilt/linux-x86_64/sysroot" \
  --cross-file android-cross.txt

# Build
ninja -C build
```

## Alternative Approach

Since mavlink-router is a C++ project using Meson, a simpler approach might be to use **Termux** on Android:

```bash
# In Termux
pkg install clang meson ninja-build git

# Then follow the standard build process
meson setup build .
ninja -C build
```

## For HereLink Specifically

HereLink uses a custom Android-based system. You might want to check if CubePilot provides a pre-built version or if you need to integrate mavlink-router into your HereLink Android image.

Hope this helps!
